src/historywork/CheckSingleLedgerHeaderWork.cpp

```diff
@@ -0,0 +1,108 @@
+    Application& app, std::shared_ptr<HistoryArchive> archive,
+    LedgerHeaderHistoryEntry const& expected)
+    : Work(app, fmt::format("check-single-ledger-header-{}",
+                            expected.header.ledgerSeq))
+    , mArchive(archive)
+    , mExpected(expected)
+{
+}
+
+CheckSingleLedgerHeaderWork::~CheckSingleLedgerHeaderWork()
+{
+}
+
+void
+CheckSingleLedgerHeaderWork::doReset()
+{
+    mGetLedgerFileWork.reset();
+    mDownloadDir =
+        std::make_unique<TmpDir>(mApp.getTmpDirManager().tmpDir(getName()));
+    uint32_t checkpoint = mApp.getHistoryManager().checkpointContainingLedger(
+        mExpected.header.ledgerSeq);
+    mFt = std::make_unique<FileTransferInfo>(
+        *mDownloadDir, HISTORY_FILE_TYPE_LEDGER, checkpoint);
+}
+
+BasicWork::State
+CheckSingleLedgerHeaderWork::doWork()
+{
+    if (!mGetLedgerFileWork)
+    {
+        CLOG_INFO(History, "Downloading ledger checkpoint {} from archive {}",
+                  mFt->baseName_gz(), mArchive->getName());
+        mGetLedgerFileWork = addWork<GetAndUnzipRemoteFileWork>(*mFt, mArchive);
+        return State::WORK_RUNNING;
+    }
+    else if (mGetLedgerFileWork->getState() != State::WORK_SUCCESS)
+    {
+        return mGetLedgerFileWork->getState();
+    }
+
+    // Theoretically we could do this scan in multiple steps with state
+    // transitions, but a ledger header file is 30kb: reading it synchronously
+    // is nearly instant.
+    XDRInputFileStream in;
+    in.open(mFt->localPath_nogz());
+    LedgerHeaderHistoryEntry lhhe;
+    try
+    {
+        while (in && in.readOne<LedgerHeaderHistoryEntry>(lhhe))
+        {
+            if (lhhe.header.ledgerSeq == mExpected.header.ledgerSeq)
+            {
+                if (lhhe == mExpected)
+                {
+                    CLOG_INFO(History,
+                              "Local ledger header {} matches archive {}",
+                              mExpected.header.ledgerSeq, mArchive->getName());
+                    return State::WORK_SUCCESS;
+                }
+                else
+                {
+                    CLOG_ERROR(
+                        History,
+                        "Local ledger header {} does not match archive {}",
+                        mExpected.header.ledgerSeq, mArchive->getName());
+                    CLOG_ERROR(History, "Expected: {}",
+                               xdr::xdr_to_string(mExpected, "Local Header"));
+                    CLOG_ERROR(History, "Found: {}",
+                               xdr::xdr_to_string(lhhe, "Archive Header"));
+                    return State::WORK_FAILURE;
+                }
+            }
+        }
+    }
+    catch (xdr::xdr_runtime_error& e)
+    {
+        CLOG_ERROR(History, "XDR error decoding {}: {}", mFt->localPath_nogz(),
+                   e.what());
+        return State::WORK_FAILURE;
+    }
+    catch (FileSystemException& e)
+    {
+        // XDRStream already logs an error on FileSystemException.
```

Comment:
do we need to give information on which archive triggered the error?

Reply:
Fixed.